lib/TPP/Dialect/Tpp/TppOps.cpp

```diff
@@ -268,34 +270,68 @@ void AddOp::getEffects(
 // GemmOp
 //===----------------------------------------------------------------------===//
 
-static bool verifyGemmShape(ShapedType shapedA, ShapedType shapedB,
-                            ShapedType shapedC) {
-  return !(shapedB.getRank() != 2 || shapedC.getRank() != 2 ||
-           shapedA.getRank() != 2);
-}
-
-static bool verifyGemmOperandsDims(ArrayRef<int64_t> shapeA,
-                                   ArrayRef<int64_t> shapeB,
-                                   ArrayRef<int64_t> shapeC) {
-  int64_t m = shapeC[0];
-  int64_t n = shapeC[1];
-  int64_t k = shapeA[1];
-  // Verify C(m, n) = A(m, k) B(k, n)
-  return !(shapeB[0] != k || shapeB[1] != n || shapeA[0] != m);
+template <typename OpTy>
+static LogicalResult
+validateVnniGemmOperand(OpTy operation, ArrayRef<int64_t> shape,
+                        Type elementType, int dimI, int dimJ) {
+  if (shape.size() == 3) {
+    if (!elementType.isBF16()) {
+      return operation->emitOpError() << "operand 1 invalid element type for "
+                                         "VNNI layout expect bf16, but got: "
+                                      << elementType << "\n";
+    }
+    if (shape[shape.size() - 1] !=
```

Comment:
No point in using `size - 1` given that the `shape.size()` is guaranteed to be 3.

Reply:
Was to avoid magic numbers.

Reply:
I thought so, but `-1` is still magic number. :)